include/forcing/DataProvider.hpp

```diff
@@ -0,0 +1,81 @@
+#ifndef NGEN_DATAPROVIDER_HPP
+#define NGEN_DATAPROVIDER_HPP
+
+#include <string>
+
+
+namespace data_access
+{
+    /** An abstraction for classes that provides access to data.
+    *
+    * Data may be pre-provided from some external source, internally calculated by the implementing type, or some
+    * combination of both.
+    */
+
+    template <class data_type, class selection_type> class DataProvider
+    {
+        /** This class provides a generic interface to data services
+        *
+        */
+
+        public:
+
+        enum ReSampleMethod
```

Comment:
I would add `front_fill` and `back_fill` as valid options here.  Not sure how we intend to offer this configurability to users, but these are two standard methods that are useful especially when working with rates over time.

Reply:
I can add the methods but then `get_data()` must handle them. What is the intended behavior?